src/runtime-rs/crates/hypervisor/src/ch/inner_device.rs

```diff
@@ -72,15 +77,37 @@ impl CloudHypervisorInner {
         match device {
             DeviceType::ShareFs(sharefs) => self.handle_share_fs_device(sharefs.config).await,
             DeviceType::HybridVsock(hvsock) => self.handle_hvsock_device(&hvsock.config).await,
-            DeviceType::Block(block) => self.handle_block_device(block.config).await,
+            DeviceType::Block(block) => self.handle_block_device(block).await,
+            DeviceType::Vfio(vfiodev) => self.handle_vfio_device(&vfiodev).await,
             _ => Err(anyhow!("unhandled device: {:?}", device)),
         }
     }
 
-    pub(crate) async fn remove_device(&mut self, _device: DeviceType) -> Result<()> {
+    /// Add the device that were requested to be added before the VMM was
+    /// started.
+    #[allow(dead_code)]
+    pub(crate) async fn handle_pending_devices_after_boot(&mut self) -> Result<()> {
+        if self.state != VmmState::VmRunning {
+            return Err(anyhow!(
+                "cannot handle pending devices with VMM state {:?}",
+                self.state
+            ));
+        }
+
+        while let Some(dev) = self.pending_devices.pop() {
+            self.add_device(dev).await.context("add_device")?;
+        }
+
         Ok(())
     }
 
+    pub(crate) async fn remove_device(&mut self, device: DeviceType) -> Result<()> {
+        match device {
+            DeviceType::Vfio(vfiodev) => self.remove_vfio_device(&vfiodev).await,
+            _ => Ok(()),
```

Comment:
Shouldn't this generate an error, or at least a log warning?

Reply:
I am returning an error here, as if device removal is not implemented, this should not error out at the moment. For certain devices, removal may not be even required, but this will have to be evaluated later on a case by case basis.